daceml/autodiff/implementations/onnx_ops.py

```diff
@@ -0,0 +1,147 @@
+        output_dtype = butils.forward_out_desc_with_name(
+            forward_node, context, "output").dtype
+
+        sums_shape = list(copy.deepcopy(output_shape))
+        sums_shape[dim] = 1
+
+        def softmax_backward(output, output_grad, input_grad):
+            prod = dace.define_local(output_shape, output_dtype)
+            sums = dace.define_local(sums_shape, output_dtype)
+            donnx.ONNXMul(A=output, B=output_grad, C=prod)
+            donnx.ONNXReduceSum(data=prod,
+                                reduced=sums,
+                                keepdims=1,
+                                axes=[dim])
+
+            donnx.ONNXMul(A=output, B=sums, C=input_grad)
+            # let's not use ONNXSub here; not sure how this inplace op is handled by ORT...
+            input_grad[:] = prod - input_grad
+
+        result_node, result = butils.backward_program_for_node(
+            softmax_backward, context, forward_node)
+
+        butils.connect_output_from_forward(forward_node, result_node, context,
+                                           "output")
+
+        return result_node, result
+
+
+@autoregister_params(op="LogSoftmax", name="default")
+class DefaultLogSoftmaxBackward(BackwardImplementation):
+    @staticmethod
+    def backward(
+        forward_node: nd.Node, context: BackwardContext,
+        given_gradients: typing.List[typing.Optional[str]],
+        required_gradients: typing.List[typing.Optional[str]]
+    ) -> typing.Tuple[nd.Node, BackwardResult]:
+
+        dim = forward_node.axis
+        output_shape = butils.forward_out_desc_with_name(
+            forward_node, context, "output").shape
+        output_dtype = butils.forward_out_desc_with_name(
+            forward_node, context, "output").dtype
+
+        sums_shape = list(copy.deepcopy(output_shape))
+        sums_shape[dim] = 1
+
+        def logsoftmax_backward(output, output_grad, input_grad):
+            exp_output = dace.define_local(output_shape, output_dtype)
+            donnx.ONNXExp(input=output, output=exp_output)
+
+            grad_output_sum = dace.define_local(sums_shape, output_dtype)
+            donnx.ONNXReduceSum(data=output_grad,
+                                reduced=grad_output_sum,
+                                keepdims=1,
+                                axes=[dim])
+            # let's not use ONNXMul here; not sure how this inplace op is handled by ORT...
+            exp_output[:] = exp_output * grad_output_sum
+            donnx.ONNXSub(A=output_grad, B=exp_output, C=input_grad)
+
+        result_node, result = butils.backward_program_for_node(
+            logsoftmax_backward, context, forward_node)
+
+        butils.connect_output_from_forward(forward_node, result_node, context,
+                                           "output")
+        return result_node, result
+
+
+@autoregister_params(op="Relu", name="pure")
+class PureReluBackward(BackwardImplementation):
```

Comment:
(no need to fix) that looks like a very long implementation for relu backward

Reply:
Yeah, looking back it would have been better to just implement a replacement for `Heaviside(x)`, but I'm fine with this for now